staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,342 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remotecommand
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	gwebsocket "github.com/gorilla/websocket"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/remotecommand"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/websocket"
+	"k8s.io/klog/v2"
+)
+
+const (
+	pingReadDeadline = 60 * time.Second
+	// should be less than pingReadDeadline
+	pingWriteDeadline = 10 * time.Second
+	PingPeriod        = 5 * time.Second
+)
+
+var (
+	_ Executor          = &wsStreamExecutor{}
+	_ streamCreator     = &wsStreamCreator{}
+	_ httpstream.Stream = &stream{}
+
+	streamType2streamID = map[string]byte{
+		v1.StreamTypeStdin:  remotecommand.StreamStdIn,
+		v1.StreamTypeStdout: remotecommand.StreamStdOut,
+		v1.StreamTypeStderr: remotecommand.StreamStdErr,
+		v1.StreamTypeError:  remotecommand.StreamErr,
+		v1.StreamTypeResize: remotecommand.StreamResize,
+	}
+)
+
+// wsStreamExecutor handles transporting standard shell streams over an httpstream connection.
+type wsStreamExecutor struct {
+	config    *restclient.Config
+	method    string
+	url       string
+	protocols []string
+}
+
+// NewWebSocketExecutor allows to execute commands via a WebSocket connection.
+func NewWebSocketExecutor(config *restclient.Config, method, url string) Executor {
+	return &wsStreamExecutor{
+		config:    config,
+		method:    method,
+		url:       url,
+		protocols: []string{remotecommand.StreamProtocolV5Name},
+	}
+}
+
+// Deprecated: use StreamWithContext instead to avoid possible resource leaks.
+// See https://github.com/kubernetes/kubernetes/pull/103177 for details.
+func (e *wsStreamExecutor) Stream(options StreamOptions) error {
+	return e.StreamWithContext(context.Background(), options)
+}
+
+func (e *wsStreamExecutor) StreamWithContext(ctx context.Context, options StreamOptions) error {
+	req, err := http.NewRequestWithContext(ctx, e.method, e.url, nil)
+	if err != nil {
+		return err
+	}
+	rt, wsRt, err := websocket.RoundTripperFor(e.config)
+	if err != nil {
+		return err
+	}
+	conn, err := websocket.Negotiate(rt, wsRt, req, e.protocols...)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	if conn == nil {
+		panic(fmt.Errorf("websocket connection is nil"))
+	}
+	var streamer streamProtocolHandler
+
+	streamingProto := conn.Subprotocol()
+	klog.V(4).Infof("The subprotocol is %s", streamingProto)
+	if streamingProto != remotecommand.StreamProtocolV5Name {
```

Comment:
only supports v5?

Reply:
This client only requests the V5remote command protocol, since it includes the stream RESET signal. No previous websocket server remote command version supports this signal. Also, there is a version skew issue. If we use this new client against a previous server (or one with the feature gate turned off), the upgrade proxy will forward to the websocket server on the container runtime (which supports V4). We do not want the websocket server on the container runtime to successfully negotiate with this client. The failure to negotiate will trigger the fallback to legacy SPDY (implemented in the . follow-on PR as `fallbackExecutor`.

Reply:
we already limit the protocols we support to v5 here: `protocols: []string{remotecommand.StreamProtocolV5Name},`

if the server doesn't support those, shouldn't Negotiate return an error already? we shouldn't need to re-check that exactly v5 subprotocol was selected here